cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java

```diff
@@ -0,0 +1,167 @@
+                                        .setNumInstances(conf.getWorkerNumNodes())
+                                        .setMachineTypeUri(conf.getWorkerMachineType())
+                                        .setDiskConfig(DiskConfig.newBuilder()
+                                                         .setBootDiskSizeGb(conf.getWorkerDiskGB())
+                                                         .setNumLocalSsds(0)
+                                                         .build())
+                                        .build())
+                     .setGceClusterConfig(GceClusterConfig.newBuilder()
+                                            .setSubnetworkUri("default")
+                                            .setZoneUri(conf.getZone())
+                                            .build())
+                     .build())
+        .build();
+
+      return client.createClusterAsync(conf.getProjectId(), conf.getRegion(), cluster)
+        .getInitialFuture().get(10, TimeUnit.SECONDS);
+    } catch (ApiException e) {
+      throw handleApiException(e);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof ApiException) {
+        throw handleApiException((ApiException) cause);
+      }
+      throw Throwables.propagate(e);
+    }  catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RetryableProvisionException(e);
+    } catch (TimeoutException e) {
+      throw new RetryableProvisionException(e);
+    }
+  }
+
+  // if there was an API exception that was not a 4xx, we can just try again
+  private RetryableProvisionException handleApiException(ApiException e) throws RetryableProvisionException {
+    if (e.getStatusCode().getCode().getHttpStatusCode() / 100 != 4) {
+      throw new RetryableProvisionException(e);
+    }
+    throw e;
+  }
+
+  @Nullable
+  public OperationSnapshot deleteCluster(String name) throws RetryableProvisionException {
+
+    try {
+      DeleteClusterRequest request = DeleteClusterRequest.newBuilder()
+        .setClusterName(name)
+        .setProjectId(conf.getProjectId())
+        .setRegion(conf.getRegion())
+        .build();
+
+      return client.deleteClusterAsync(request).getInitialFuture().get(10, TimeUnit.SECONDS);
+    } catch (ApiException e) {
+      if (e.getStatusCode().getCode().getHttpStatusCode() == 404) {
+        // if the cluster was not found, it's ok that means it's deleted
+        return null;
+      }
+      throw handleApiException(e);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof ApiException) {
+        throw handleApiException((ApiException) cause);
+      }
+      throw Throwables.propagate(e);
+    }  catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RetryableProvisionException(e);
+    } catch (TimeoutException e) {
+      throw new RetryableProvisionException(e);
```

Comment:
Is it safe to issue the `deleteClusterAsync` again on timeout of the `getInitialFuture.get` call? Why can't the `get` block indefinitely?

Reply:
hm I suppose it's not correct. For the timeout, I was worried about environments where these public APIs are blocked from the network and these requests hang forever, but in those situations the request should not be retried.